[3508] current_user support for multiple organisations

gpeng opened the pull request:
### Context

Users can now be associated with any combination of multiple providers or lead schools. The associations have been created but we're still not supporting this in the UI. Ultimately when a user logs in they should either see the correct context (either lead school or provider) if they have only one organisation or be given the opportunity to choose which context they want to work in.

### Changes proposed in this pull request

* Add support for setting the current organisation context and passing that around with current_user
* Add a multi org persona for testing
* Add a feature flag to disable this in production until it's complete

There is still work to do with this before it is ready. Still to do:

* lead school view should be read only. The policy has been updated but we're not hiding any of the change links etc
* a component to allow the user to switch their organisation context
* back link and nav hiding on the organisations#index page when a user first logs in
* some inevitable snagging



### Guidance to review

Most importantly at this stage we're checking that the organisation context displays the correct trainees for that organisation. Also that when we turn the feature flag off everything still works and the multi-org stuff doesn't kick in anywhere.

The TraineePolicy has gotten a little bit out of hand as we've mixed concerns a bit and included trainee state transition logic. This has made the tests a bit sprawling. We should maybe address that in the future but I've left it for now.

Removing `current_user.primary_provider` has caused 'a lot ™️' of changes. I've tried to keep those contained in one commit. Might be best to review one commit at a time. Sorry 😬 

<img width="1063" alt="Screenshot 2022-01-28 at 17 28 23" src="https://user-images.githubusercontent.com/5216/151593748-4b6317e9-9850-4846-a872-2c7c842a2574.png">

### Important business

* ~Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?~
* ~Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?~

jhenderson reviewed (changes requested):
Looks good. 👍  Left some minor feedback and request changes.

gpeng commented on app/controllers/system_admin/users_controller.rb:
```diff
@@ -22,17 +22,16 @@ def create
     def edit
       user
-      @provider = user.primary_provider
+      provider
```
Didn't we decide a while back to prefer memoized accessors over before actions? We've not been super consistent with it though.

gpeng commented on app/controllers/pages_controller.rb:
```diff
@@ -6,6 +6,10 @@ class PagesController < ApplicationController
   def start
     session[:requested_path] = root_path
     if authenticated?
+      if FeatureService.enabled?(:user_can_have_multiple_organisations) && (current_user.organisation.nil? && !current_user.system_admin?)
```
👍

gpeng commented on app/models/dfe_sign_in_user.rb:
```diff
@@ -48,7 +48,7 @@ def self.load_from_session(session)
 
   def self.end_session!(session)
-    session.delete("dfe_sign_in_user")
+    session.destroy
```
We've got some extra keys in there, particularly the organisation context one, that we want to clear when the user logs out so I figured we should just clear the whole session. Without this when the user logs back in they don't get the organisation picker as the context is still set from the previous session.

gpeng commented:
Thanks @jhenderson I've addressed your feedback in the fixup commits (I'll rebase them in when you're happy with 'em). Still need to figure out how to make sonarcloud go away though 🤫

mooktakim commented:
![](https://github.trello.services/images/mini-trello-icon.png) [current_user support for multiple organisations](https://trello.com/c/NNKrZru7/3508-currentuser-support-for-multiple-organisations)

jhenderson commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[S] Main navigation should be hidden when a user has no organisation context set](https://trello.com/c/YyI7oCis/3602-s-main-navigation-should-be-hidden-when-a-user-has-no-organisation-context-set)
